recipes_source/recipes/benchmark.py

```diff
@@ -0,0 +1,828 @@
+            num_threads=num_threads,
+            label="Shift operator",
+            sub_label="Generic implementation.",
+            description=str(n),
+        ).blocked_autorange())
+
+    results.append(Timer(
+        "my_module.shift(x)",
+        setup=(
+            module_to_setup_str(shift_impl_v0) +
+            setup
+        ),
+        # Custom C++ operator does not support parallelism.
+        num_threads=1,
+        label="Shift operator",
+        sub_label="Custom C++ operator",
+        description=str(n),
+    ).blocked_autorange())
+
+compare = Compare(results)
+compare.print()
+
+# [------------------------------------- Shift operator ------------------------------------]
+#                                |   1   |   16  |  256  |  1024  |  4096  |  16384  |  32768
+# 1 threads: --------------------------------------------------------------------------------
+#       Generic implementation.  |  8.2  |  8.6  |  8.7  |  9.3   |  10.3  |   14.7  |   22.3
+#       Custom C++ operator      |  4.1  |  4.2  |  4.5  |  5.3   |   8.2  |   18.6  |   35.3
+# 2 threads: --------------------------------------------------------------------------------
+#       Generic implementation.  |  8.0  |  8.4  |  8.7  |  9.3   |  10.4  |   14.6  |   23.4
+# 4 threads: --------------------------------------------------------------------------------
+#       Generic implementation.  |  7.9  |  8.7  |  8.9  |  9.3   |  10.5  |   14.6  |   23.8
+#
+# Times are in microseconds (us).
+
+######################################################################
+# With extra formatting (colors might not be shown here).
+#
+
+compare.trim_significant_figures()
+compare.colorize()
+compare.print()
+
+# [------------------------------------- Shift operator ------------------------------------]
+#                                |   1   |   16  |  256  |  1024  |  4096  |  16384  |  32768
+# 1 threads: --------------------------------------------------------------------------------
+#       Generic implementation.  |  8.2  |  8.6  |  8.7  |  9.3   |   10   |    15   |    22 
+#       Custom C++ operator      |  4.1  |  4.2  |  4.5  |  5.3   |    8   |    19   |    35 
+# 2 threads: --------------------------------------------------------------------------------
+#       Generic implementation.  |  8.0  |  8.4  |  8.7  |  9.3   |   10   |    15   |    23 
+# 4 threads: --------------------------------------------------------------------------------
+#       Generic implementation.  |  7.9  |  8.7  |  8.9  |  9.3   |   11   |    15   |    24 
+#
+# Times are in microseconds (us).
+
+######################################################################
+# 6. Generating bechmark inputs with Fuzzed Parameters
+#
+# We'll take a brief detour and use fuzzed inputs to discuss transpose
+# and contiguous before returning to shift.
+#
+
+from torch.utils.benchmark import Fuzzer, FuzzedParameter, FuzzedTensor, ParameterAlias
+
+example_fuzzer = Fuzzer(
```

Comment:
a description of what this fuzzer does (generates 2d tensors with sizes distributed according to specified distributions) is useful. You are talking to people who see Fuzzer and FuzzedParameter for the first time.